desc/optimize/optimizer.py

```diff
@@ -543,3 +343,108 @@ def _get_default_tols(
     stoptol["max_nhev"] = options.pop("max_nhev", np.inf)
 
     return stoptol
+
+
+optimizers = {}
+
+
+def register_optimizer(
+    name,
+    scalar,
+    equality_constraints,
+    inequality_constraints,
+    stochastic,
+    hessian,
+    **kwargs,
+):
+    """Decorator to wrap a function for optimization.
+
+    Function being wrapped should have a signature of the form
+    fun(objective, constraint, x0, method, x_scale, verbose, stoptol, options=None)
+    and should return a scipy.optimize.OptimizeResult object
+
+    Function should take the following arguments:
+
+    objective : ObjectiveFunction
+        Function to minimize.
+    constraint : ObjectiveFunction
+        Constraint to satisfy
+    x0 : ndarray
+        Starting point.
+    method : str
+        Name of the sub-method to use.
+    x_scale : array_like or ‘jac’, optional
+        Characteristic scale of each variable.
+    verbose : int
+        * 0  : work silently.
+        * 1 : display a termination report.
+        * 2 : display progress during iterations
+    stoptol : dict
+        Dictionary of stopping tolerances, with keys {"xtol", "ftol", "gtol",
+        "maxiter", "max_nfev", "max_njev", "max_ngev", "max_nhev"}
+    options : dict, optional
+        Dictionary of optional keyword arguments to override default solver
+        settings.
+
+
+    Parameters
+    ----------
+    name : str or array-like of str
+        Name of the optimizer method. If one function supports multiple methods,
+        provide a list of names.
+    scalar : bool or array-like of bool
+        Whether the method assumes a scalar residual, or a vector of residuals for
+        least squares.
+    equality_constraints : bool or array-like of bool
+        Whether the method handles equality constraints.
+    inequality_constraints : bool or array-like of bool
+        Whether the method handles inequality constraints.
+    stochastic : bool or array-like of bool
+        Whether the method can handle noisy objectives.
+    hessian : bool or array-like of bool
+        Whether the method requires calculation of the full hessian matrix.
+    """
+    (
+        name,
+        scalar,
+        equality_constraints,
+        inequality_constraints,
+        stochastic,
+        hessian,
+    ) = map(
+        np.atleast_1d,
+        (
+            name,
+            scalar,
+            equality_constraints,
+            inequality_constraints,
+            stochastic,
+            hessian,
+        ),
+    )
+    (
+        name,
+        scalar,
+        equality_constraints,
+        inequality_constraints,
+        stochastic,
+        hessian,
+    ) = np.broadcast_arrays(
```

Comment:
https://numpy.org/doc/stable/reference/generated/numpy.broadcast_arrays.html
is this function deprecated?

Reply:
No, just writing to the broadcasted arrays is deprecated, but we don't do that, its just to get a view of everything with the same length for iteration.